### PR TITLE
feat: Add credential provider utility classes for AWS, GCP

### PR DESCRIPTION
**Pull request by nameexhaustion**
ref https://github.com/pola-rs/polars/pull/19271#discussion_r1804601805

* Adds `CredentialProviderAWS`, `CredentialProviderGCP`

`CredentialProviderAWS` can be used to select a different AWS profile:

```python
lf = pl.scan_parquet(
    "s3://...",
    credential_provider=pl.CredentialProviderAWS(profile_name="..."),
)

# `CredentialProviderAWS()` will also pick up `AWS_PROFILE` from the environment
import os
os.environ['AWS_PROFILE'] = 'dummy'

lf = pl.scan_parquet(
    "s3://...",
    credential_provider=pl.CredentialProviderAWS(),
)
```

Todos:
* In local testing, GCP is not being cached properly

### CODE REVIEW

File: crates/polars-io/src/cloud/credential_provider.rs [hunk 1]
```diff
@@ -625,20 +632,8 @@ mod python_impl {
         }
     }
 
-    impl Eq for PythonCredentialProvider {}
-
-    impl PartialEq for PythonCredentialProvider {
-        fn eq(&self, other: &Self) -> bool {
-            Arc::ptr_eq(&self.0, &other.0)
```
Comment (nameexhaustion) [inline crates/polars-io/src/cloud/credential_provider.rs:632]: Removed, completely unsound as a new object could be allocated in the same place

Comment (ritchie46) [reply crates/polars-io/src/cloud/credential_provider.rs:632]: Don't we keep that object alive?

File: py-polars/polars/io/parquet/functions.py [hunk 2]
```diff
@@ -338,7 +336,7 @@ def scan_parquet(
     low_memory: bool = False,
     cache: bool = True,
     storage_options: dict[str, Any] | None = None,
-    credential_provider: CredentialProviderFunction | None = None,
+    credential_provider: CredentialProviderFunction | Literal["auto"] | None = "auto",
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply py-polars/polars/io/parquet/functions.py:339]: Alright. :+1:

[the remark this answers is not in the document]
Comment (nameexhaustion) [reply py-polars/polars/io/parquet/functions.py]: changed default to None

**Comment by nameexhaustion** [comment]
I did some testing and it turns out object store already does the default behaviors of `CredentialProviderAWS` and `CredentialProviderGCP` for us (i.e. loading the default credentails), so maybe we don't actually need the "auto" behavior.

`CredentialProviderGCP` also currently doesn't provide any extra functionality over object store. Although `CredentialProviderAWS` can be used to select a different AWS profile.

File: crates/polars-io/src/cloud/credential_provider.rs (hunk 1 above)
Comment (nameexhaustion) [reply crates/polars-io/src/cloud/credential_provider.rs]: Ah, you're right - I got a bit confused in my head 😅

File: py-polars/docs/source/reference/io.rst [hunk 3]
```diff
@@ -117,3 +117,14 @@ Connect to pyarrow datasets.
    :toctree: api/
 
    scan_pyarrow_dataset
+
+Cloud Credentials
+~~~~~~~~~~~~~~~~~
+Configuration for cloud credential provisioning.
```
Comment (nameexhaustion) [inline py-polars/docs/source/reference/io.rst:123]: Also added to the Python user guide

**Comment by ritchie46** [comment]
Nice, we should quickly setup an azure env as well.

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**